keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+part of the resource.k8s.io API. Therefore developers who want to extend
+a scheduler with some additional model have to modify the apiserver.
+
+When opting into using semantic parameters for their DRA driver, driver
+developers no longer need to provide a control plane controller which
+handles claim allocation. Instead, they need to provide a CRD controller
+which translates from their custom claim parameters CRD into the parameter
+object understood by Kubernetes.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
+  define configuration of resources, but must not affect where a claim can be
+  allocated.
+
+- Define exactly one semantic model that covers all current and future resources.
+  There cannot be a "one size fits all" solution.
+
+## Proposal
+
+Each node publishes information about the nearly-opaque resources it currently
+has. Each driver manages one or more resource instance per node. An instance
+could be a single discreet piece of hardware like a PCI card or a pool of such
+cards. Each instance has certain attributes. These can be qualitative (labels
+or fields for comparison) and quantitative (capacity that gets reduced when
+allocated for a claim). These attributes are described by the "attributes type"
+of the semantic models. Each instance can combine attributes from different
+semantic models. This way, each model can focus on one aspect, like counting
+items or selection via labels.
```

Comment:
I don't quite understand how a `capacity` can be thought to live directly alongside the rest of an instance's attributes. I would think there are a set of attributes that **_define_** what the instance represents, and then a separate `capacity` (in some sort of wrapping struct) to define how many of those instances are available for allocation.

Something like the following (I know the types for the attributes would need to be generalized, but you get the idea):
```
Instance {
  Driver: "gpu.nvidia.com",
  Name: "gpu",
  Attributes: {
	MigCapable            bool
	Memory                Quantity
	ProductName           string
	Brand                 string
	Architecture          string
	CUDAComputeCapability string
	DriverVersion         SemVer
	CUDARuntimeVersion    SemVer
  }
}

InstanceCapacity{
  Driver: "gpu.nvidia.com",
  Name: "gpu",
  Capacity: 8,
}
```

Reply:
I agree - I have not yet sent all my comments, but we do need to think about what we are modeling, and maybe not FULLY denormalize it?